packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java

```diff
@@ -0,0 +1,142 @@
+
+  public PreviewHostApiImpl(
+      BinaryMessenger binaryMessenger,
+      InstanceManager instanceManager,
+      TextureRegistry textureRegistry) {
+    this.binaryMessenger = binaryMessenger;
+    this.instanceManager = instanceManager;
+    this.textureRegistry = textureRegistry;
+  }
+
+  /** Creates a {@link Preview} with the target rotation and resolution if specified. */
+  @Override
+  public void create(
+      @NonNull Long identifier,
+      @Nullable Long rotation,
+      @Nullable GeneratedCameraXLibrary.ResolutionInfo targetResolution) {
+    Preview.Builder previewBuilder = cameraXProxy.createPreviewBuilder();
+    if (rotation != null) {
+      previewBuilder.setTargetRotation(rotation.intValue());
+    }
+    if (targetResolution != null) {
+      previewBuilder.setTargetResolution(
+          new Size(
+              targetResolution.getWidth().intValue(), targetResolution.getHeight().intValue()));
+    }
+    Preview preview = previewBuilder.build();
+    instanceManager.addDartCreatedInstance(preview, identifier);
+  }
+
+  /**
+   * Sets the {@link Preview.SurfaceProvider} that will be used to provide a {@code Surface} backed
+   * by a Flutter {@link TextureRegistry.SurfaceTextureEntry} used to build the {@link Preview}.
+   */
+  @Override
+  public Long setSurfaceProvider(@NonNull Long identifier) {
+    Preview preview = (Preview) instanceManager.getInstance(identifier);
+    flutterSurfaceTexture = textureRegistry.createSurfaceTexture();
+    SurfaceTexture surfaceTexture = flutterSurfaceTexture.surfaceTexture();
+    Preview.SurfaceProvider surfaceProvider =
+        new Preview.SurfaceProvider() {
+          @Override
+          public void onSurfaceRequested(SurfaceRequest request) {
+            surfaceTexture.setDefaultBufferSize(
+                request.getResolution().getWidth(), request.getResolution().getHeight());
+            Surface flutterSurface = cameraXProxy.createSurface(surfaceTexture);
+            request.provideSurface(
+                flutterSurface,
+                Executors.newSingleThreadExecutor(),
+                (result) -> {
+                  SystemServicesFlutterApiImpl systemServicesFlutterApi =
+                      cameraXProxy.createSystemServicesFlutterApi(binaryMessenger);
+                  int resultCode = result.getResultCode();
+                  switch (resultCode) {
+                    case SurfaceRequest.Result.RESULT_SURFACE_USED_SUCCESSFULLY:
+                      flutterSurface.release();
+                      break;
+                    case SurfaceRequest.Result.RESULT_REQUEST_CANCELLED:
+                    case SurfaceRequest.Result.RESULT_INVALID_SURFACE:
+                    case SurfaceRequest.Result.RESULT_WILL_NOT_PROVIDE_SURFACE:
+                      flutterSurface.release();
+                    case SurfaceRequest.Result.RESULT_SURFACE_ALREADY_PROVIDED:
+                    default:
```

Comment:
This switch is confusing. 
Success releases the surface and breaks but so does canceled invalid and "will not provide". 
Canceled feels like an expected condition but invalid and will not provide feel like errors even though they fall through and have the same behavior as success. 
Finally Already provided feels like a code error but it falls through to default which is usually dangerous. 


I am not sure what the expected conditions are but I might suggest trying to structure this in a way that is a bit easier to follow. 
For example if the result code is a actually being used as a proxy for if we have a flutterSurface then maybe we always check flutterSurface and if it exists it is released. Then if we get a value that should return an error we do that. Finally we should ensure we dont hit default and if we do perhaps we error and then test that all values of Result code do not error in a unit test.

Reply:
Modified this and replied below: https://github.com/flutter/plugins/pull/7046#discussion_r1093862234